source/module_io/write_libxc_r.cpp

```diff
@@ -0,0 +1,285 @@
+			case 3:		v3rho2sigma .resize( nrxx * ((1==nspin)?1:9)  );
+						v3rhosigma2 .resize( nrxx * ((1==nspin)?1:12) );
+						v3sigma3    .resize( nrxx * ((1==nspin)?1:10) );
+			case 2:		v2rhosigma  .resize( nrxx * ((1==nspin)?1:6)  );
+						v2sigma2    .resize( nrxx * ((1==nspin)?1:6)  );
+			case 1:		vsigma      .resize( nrxx * ((1==nspin)?1:3)  );
+			case 0:		break;
+			default:	throw std::domain_error("order ="+std::to_string(order)
+							+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+						break;
+		}
+	}
+
+	for( xc_func_type &func : funcs )
+	{
+		// jiyy add for threshold
+		constexpr double rho_threshold = 1E-6;
+		constexpr double grho_threshold = 1E-10;
+
+		xc_func_set_dens_threshold(&func, rho_threshold);
+
+		// sgn for threshold mask
+		const std::vector<double> sgn = XC_Functional_Libxc::cal_sgn(rho_threshold, grho_threshold, func, nspin, nrxx, rho, sigma);
+
+		// call libxc function
+		// attention: order 432 don't break
+		switch( func.info->family )
+		{
+			case XC_FAMILY_LDA:
+			{
+				switch( order )
+				{
+					case 4:		xc_lda_lxc    ( &func, nrxx, rho.data(), v4rho4.data() );
+					case 3:		xc_lda_kxc    ( &func, nrxx, rho.data(), v3rho3.data() );
+					case 2:		xc_lda_fxc    ( &func, nrxx, rho.data(), v2rho2.data() );
+					case 1:		xc_lda_exc_vxc( &func, nrxx, rho.data(), exc.data(), vrho.data() );
+								break;
+					case 0:		xc_lda_exc    ( &func, nrxx, rho.data(), exc.data() );
+								break;
+					default:	throw std::domain_error("order ="+std::to_string(order)
+									+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+								break;
+				}
+				break;
+			}
+			case XC_FAMILY_GGA:
+			case XC_FAMILY_HYB_GGA:
+			{
+				switch( order )
+				{
+					case 4:		xc_gga_lxc    ( &func, nrxx, rho.data(), sigma.data(), v4rho4.data(), v4rho3sigma.data(), v4rho2sigma2.data(), v4rhosigma3.data(), v4sigma4.data() );
+					case 3:		xc_gga_kxc    ( &func, nrxx, rho.data(), sigma.data(), v3rho3.data(), v3rho2sigma.data(), v3rhosigma2.data(), v3sigma3.data() );
+					case 2:		xc_gga_fxc    ( &func, nrxx, rho.data(), sigma.data(), v2rho2.data(), v2rhosigma.data(), v2sigma2.data() );
+					case 1:		xc_gga_exc_vxc( &func, nrxx, rho.data(), sigma.data(), exc.data(), vrho.data(), vsigma.data() );
+								break;
+					case 0:		xc_gga_exc    ( &func, nrxx, rho.data(), sigma.data(), exc.data() );
+								break;
+					default:	throw std::domain_error("order ="+std::to_string(order)
+									+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+								break;
+				}
+				break;
+			}
+			default:
+			{
+				throw std::domain_error("func.info->family ="+std::to_string(func.info->family)
+					+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+				break;
+			}
+		} // end switch( func.info->family )
+	} // end for( xc_func_type &func : funcs )
+
+	auto write_data = [&esolver](

```

Comment:
I think it's better to let `write_data` be a free function ( named like `write_xc_r`) instead of a lambda in such a big function `write_libxc_r`. It will help someone like me already having data like `fxc(nrxx, 3)` calculated elsewhere to write.
The dependences can be pw_big, pw_rhod, nspin, ld(number_spin), directory and precision, independent of `PARAM` and `esolver`.